backend/src/main/java/com/javahelp/backend/domain/user/register/UserRegisterInteractor.java

```diff
@@ -0,0 +1,80 @@
+        U extends UserInfo>{
+
+    IUserStore userStore;
+
+    /**
+     * Creates a new {@link UserRegisterInteractor} instance
+     *
+     * @param userStore  {@link IUserStore} to use
+     */
+    public UserRegisterInteractor(IUserStore userStore) {
+        this.userStore = userStore;
+    }
+
+    /**
+     * Checking for violations for registering any {@link User} account.
+     * @param boundary an instance of {@link IUserRegisterInputBoundary} to retrieve input info
+     *
+     * @return a {@link RegisterUserResponse} model that contains message of failure at registration
+     * Null if all tests passed.
+     */
+    protected RegisterUserResponse areEmailAndUsernameTaken(T boundary){
+        if ((userStore.readByEmail(boundary.getEmailAddress()) != null)
+                && (userStore.readByUsername(boundary.getUsername()) != null)){
+            // If both username and email address exist, then return both errors
+            return new RegisterUserResponse("Both the given email address and username " +
+                    "are already occupied");
+        }
+        if (IUserStore.getDefaultImplementation().readByEmail(boundary.getEmailAddress()) != null){
+            return new RegisterUserResponse("The email address is already occupied");
+        }
+        else if (IUserStore.getDefaultImplementation().readByUsername(boundary.getUsername()) != null){
+            return new RegisterUserResponse("The username is already occupied.");
+        }
+
+        // Nothing goes wrong
+        return null;
+    }
+
+    /**
+     * A method for future different types of {@link User} that might need to validate
+     * their own defined attributes (i.e. beyond the uniqueness check in username and email address)
+     * @param boundary <T> interface extends {@link IUserRegisterInputBoundary} to retrieve input
+     *                 info
+     *
+     * @return a {@link RegisterUserResponse} model that has a null error message if and only if
+     * the input passed all tests.
+     */
+    protected abstract RegisterUserResponse isValidated(T boundary);
+
+    /**
+     * A method to bundle up {@link UserInfo} based on their given
+     * {@link com.javahelp.model.user.UserType}
+     *
+     * @return a {@link UserInfo} instance that contains all input data
+     */
+    protected abstract U bundleUserInfo(T boundary);
+
+    /**
+     * An abstract method to create a new user entity for this user based on their info,
+     * which will be defined in other subtypes of users.
+     * @param boundary <T> interface extends {@link IUserRegisterInputBoundary} to get info from the
+     *                 register View Model.
+     *
+     * @return a new {@link RegisterUserResponse} entity that comprises of {@link User}
+     * {@link UserPassword} if success
+     * OR of only a {@link String} error message if failed.
+     */
+    protected abstract RegisterUserResponse register(T boundary);
```

Comment:
This method should not be abstract, it has a very well defined behaviour, that will be the same in every implementation. To pass the created `User` object, with the `UserInfo` created in `bundleUserInfo` along with the created `UserPassword` object to `userStore.create`. This behaviour is the same for every single implementation so making this abstract doesn't make much sense.